File: examples/Exceptions.py
#!/usr/bin/env python3

# pylint: disable = bad-whitespace
# pylint: disable = invalid-name
# pylint: disable = line-too-long
# pylint: disable = literal-comparison
# pylint: disable = missing-docstring
# pylint: disable = unidiomatic-typecheck

# -------------
# Exceptions.py
# -------------

# https://docs.python.org/3/library/exceptions.html

def f (b) :
    if b :
        raise NameError("abc")
    return 0

def test1 () :
    try :
        assert f(False) == 0
    except NameError :       # pragma: no cover
        assert False
    else :                   # no exception vs. raised and handled, but not with break, continue, return
        pass
    finally :                # always, even with break, continue, return
        pass

def test2 () :
    try :
        assert f(True) == 1
        assert False                         # pragma: no cover
    except NameError as e :
        assert type(e)      is NameError
        assert type(e.args) is tuple
        assert len(e.args)  ==     1
        assert e.args       is not ("abc",)
        assert e.args       ==     ("abc",)
    else :                                   # no exception vs. raised and handled, but not with break, continue, return
        assert False                         # pragma: no cover
    finally :                                # always, even with break, continue, return
        pass

    assert issubclass(NameError,     Exception)
    assert issubclass(Exception,     BaseException)
    assert issubclass(BaseException, object)

if __name__ == "__main__" : # pragma: no cover
    print("Exceptions.py")
    test1()
    test2()
    print("Done.")

File: examples/Exceptions.js
#!/usr/bin/env node

/*global
    require
*/

/*jslint
    white: true
*/

// -------------
// Exceptions.js
// -------------

// https://nodejs.org/api/errors.html

"use strict";

var assert = require('assert');

function f (b) {
    if (b) {
        throw new Error("abc");}
    return 0;}

function test1 () {
    try {
        assert(f(false) === 0);}
    catch (e) {
        assert(false);}
    finally {
        assert(true);}}

function test2 () {
    try {
        assert(f(true) === 1);
        assert(false);}
    catch (e) {
        assert(e         instanceof Error);
        assert(e.message ===        "abc");}
    finally {
        assert(true);}}

console.log("Exceptions.js");
test1();
test2();
console.log("Done.");
